pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,394 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"path"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+// bitsDataElement is the number of bits in a bitArray.data element.
+const bitsDataElement = 32
+
+type bitArray struct {
+	data       []uint32
+	firstIndex int
+}
+
+func makeBitArray(size uint32) *bitArray {
+	m := bitArray{
+		data:       make([]uint32, (size+bitsDataElement-1)/bitsDataElement),
+		firstIndex: 0,
+	}
+	return &m
+}
+
+func (b *bitArray) set(index uint32) {
+	b.data[index/bitsDataElement] |= (uint32(1) << (index % bitsDataElement))
+}
+
+func (b *bitArray) isSet(index uint32) bool {
+	return (b.data[index/bitsDataElement]>>(index%bitsDataElement))&0x1 == 1
+}
+
+func (b *bitArray) findAndSetFirstZero() (uint32, error) {
+	for i := b.firstIndex; i < len(b.data); i++ {
+		if b.data[i] == 0xFFFFFFFF {
+			continue
+		}
+		for j := uint32(0); j < bitsDataElement; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsDataElement + j
+				b.set(v)
+				b.firstIndex = int(i)
+				return v, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not find an empty slot")
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsDataElement
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	b.data[i] &= ^(1 << (index % bitsDataElement))
+}
+
+// length for the user namespace to create (enough to fit 16 bits IDs).
+const userNsLength = (1 << 16)
+
+// Limit the total number of pods using userns in this node to this value.
+// This is an alpha limitation that will probably be lifted later.
+const maxPods = 1024
```

Comment:
Should this be tied to MaxPods validation when the feature flag is enabled?

Reply:
this limit is specified in the KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/127-user-namespaces/README.md#phase-1-pods-without-volumes:

```
During phase 1, to make sure we don't exhaust the host UID namespace, we will limit the number of pods using user namespaces to min(maxPods, 1024). This leaves us plenty of host UID space free and this limits is probably never hit in practice. See UNRESOLVED for more some UNRESOLVED info we still have on this.
```

Personally I think there should not be any such artificial limitation and use all the available IDs.

If we want to restrict the IDs that are usable for user namespaces, then that is a different feature, which should be discussed separately

Reply:
@giuseppe Agree :)

@mrunalp also, what do you mean to tie this to MaxPods in validation? We want this limit per node, not per cluster, so we can't do it at API validation IIUC.

Also, the only say I can think of tying them up (at kubelet pod creation time) is to do something like:

```
 usernsMaxPods := min(kubeletMaxPods, 1024)
```

But that is the same as what we achieve with this code. If KubeletMaxPods is lower than 1024, then the kubelet will not create more pods and these functions will not even be called. If kubeletMaxPods is higher than usernsMaxPods, then this limit as specified in the KEP is used.

Reply:
kubelet has `--max-pods ` that specifies the number of Pods that can run on this Kubelet [1], which is 110 by default. I guess this is what mrunalp talks about. 

[1] https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

Reply:
if i set --max-pods to greater than this value, and all pods scheduled used user namespace feature, what happens when we cannot allocate a user namespace for this pod?  the pod will remain pending i assume until a slot is freed?

Reply:
@derekwaynecarr sorry, missed this comment yesterday. Yes, this is exactly what happens. See here for example output too: https://github.com/kubernetes/kubernetes/pull/111090#discussion_r925737543

Reply:
> Personally I think there should not be any such artificial limitation and use all the available IDs. 

When we get to a later phase, we'll regret that.  We need some way to allocate IDs deterministically for stateful pods, and if we make a mess now, that will be harder.